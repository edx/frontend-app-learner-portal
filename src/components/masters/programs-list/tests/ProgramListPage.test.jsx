import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { StaticQuery } from 'gatsby';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { StatusAlert } from '@edx/paragon';

import ConnectedProgramListPage, { ProgramListPage } from '../ProgramListPage';

const mockStore = configureMockStore([thunk]);

describe('ProgramListPage', () => {
  beforeEach(() => {
    StaticQuery.mockImplementationOnce(({ render }) => (
      render({
        site: {
          siteMetadata: {
            logo: 'https://edx.org',
            siteName: 'edX',
            siteUrl: 'https://edx.org',
          },
        },
      })
    ));
  });

  const programQueryData = [
    {
      node: {
        context: {
          programUUID: '6eefc008-db50-46f0-8746-667f55533a5d',
          programName: 'Example Program',
          programSlug: 'exampleprogram',
          programHostname: 'http://localhost:8734',
        },
      },
    },
    {
      node: {
        context: {
          programUUID: '6eefc008-db50-46f0-8746-667f55533a5e',
          programName: 'Another Program',
          programSlug: 'another-program',
          programHostname: 'http://localhost:8734/another-program',
        },
      },
    }];

  it('correctly renders the loading page', () => {
    const store = mockStore({
      authentication: {
        username: 'edx',
      },
      userAccount: {
        profileImage: {
          imageUrlMedium: 'someImageData',
        },
      },
      enrolledPrograms: {
        loading: true,
        error: null,
      },
    });

    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <Provider store={store}>
            <ConnectedProgramListPage programQueryData={programQueryData} />
          </Provider>
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders fetching program error page when there are issues fetching the user programs', () => {
    const store = mockStore({
      authentication: {
        username: 'edx',
      },
      userAccount: {
        profileImage: {
          imageUrlMedium: 'someImageData',
        },
      },
      enrolledPrograms: {
        loading: false,
        error: new Error(),
      },
    });

    const tree = renderer
      .create((
        <IntlProvider locale="en">
          <Provider store={store}>
            <ConnectedProgramListPage programQueryData={programQueryData} />
          </Provider>
        </IntlProvider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders table when there are valid programs', () => {
    const data = [
      {
        uuid: '6eefc008-db50-46f0-8746-667f55533a5d',
        name: 'Example Program',
        slug: 'exampleprogram',
      },
      {
        uuid: '6eefc008-db50-46f0-8746-667f55533a5e',
        name: 'Another Program',
        slug: 'another-program',
      },
    ];

    const wrapper = shallow((
      <ProgramListPage
        store={mockStore()}
        programQueryData={programQueryData}
        isLoading={false}
        fetchUserProgramEnrollments={jest.fn()}
      />
    ));

    wrapper.setProps({ enrolledPrograms: data });

    expect(wrapper.find('.table-responsive').exists()).toBeTruthy();
    expect(wrapper.find('tbody tr').length).toEqual(2);
  });

  it('renders error page when there are no valid programs', () => {
    const data = [
      {
        uuid: 'this-is-a-fake-program',
        name: 'This is a fake program',
        slug: 'fake-program',
      },
      {
        uuid: 'this-is-another-fake-program',
        name: 'This is another fake program',
        slug: 'another-fake-program',
      },
    ];

    const wrapper = shallow((
      <ProgramListPage
        store={mockStore()}
        programQueryData={programQueryData}
        isLoading={false}
        fetchUserProgramEnrollments={jest.fn()}
      />
    ));

    wrapper.setProps({ enrolledPrograms: data });

    expect(wrapper.find('.table-responsive').exists()).toBeFalsy();
    expect(wrapper.find(StatusAlert).exists()).toBeTruthy();
  });
});
