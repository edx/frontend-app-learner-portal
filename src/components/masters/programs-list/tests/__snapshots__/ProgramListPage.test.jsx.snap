// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgramListPage correctly renders the loading page 1`] = `
Array [
  <div
    className="position-absolute"
  >
    <a
      className="skip-nav-link sr-only sr-only-focusable btn btn-primary px-2 py-1 mt-3 ml-2"
      href="#content"
    >
      Skip to main content
    </a>
  </div>,
  <div
    className="container my-4"
  >
    <div
      className="d-flex justify-content-center align-items-center"
      style={
        Object {
          "height": 200,
        }
      }
    >
      <div
        className="spinner-border text-primary"
        role="status"
      >
        <div
          className="sr-only"
        >
          Loading program enrollments...
        </div>
      </div>
    </div>
  </div>,
  <footer
    aria-label="Page Footer"
    className="footer d-flex justify-content-center border-top py-3 px-4"
    role="contentinfo"
  >
    <div
      className="max-width-1180 d-grid"
    >
      <div
        className="area-1"
      >
        <a
          aria-label="Open Edx Home"
          href="https://www.example.com/"
          onClick={[Function]}
          target="_self"
        >
          <img
            alt="Open Edx logo"
            src={
              Object {
                "process": [Function],
              }
            }
          />
        </a>
      </div>
      <div
        className="area-2"
      >
        <h2>
          Open Edx
        </h2>
        <ul
          className="list-unstyled p-0 m-0"
        >
          <li>
            <a
              href="https://www.example.com/about-us"
            >
              <span>
                About
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/enterprise"
            >
              <span>
                Open Edx for Business
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/affiliate-program"
            >
              <span>
                Affiliates
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/open"
            >
              <span>
                Open Open Edx
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/careers"
            >
              <span>
                Careers
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/news-announcements"
            >
              <span>
                News
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        className="area-3"
      >
        <h2>
          <span>
            Legal
          </span>
        </h2>
        <ul
          className="list-unstyled p-0 m-0"
        >
          <li>
            <a
              href="https://www.example.com/terms-of-service"
            >
              <span>
                Terms of Service & Honor Code
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/privacy-policy"
            >
              <span>
                Privacy Policy
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/accessibility"
            >
              <span>
                Accessibility Policy
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/trademarks"
            >
              <span>
                Trademark Policy
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/sitemap"
            >
              <span>
                Sitemap
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        className="area-4"
      >
        <h2>
          <span>
            Connect
          </span>
        </h2>
        <ul
          className="list-unstyled p-0 m-0"
        >
          <li>
            <a
              href="https://www.example.com/blog"
            >
              <span>
                Blog
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/contact"
            >
              <span>
                Contact Us
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/support"
            >
              <span>
                Help Center
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/media-kit"
            >
              <span>
                Media Kit
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/donate"
            >
              <span>
                Donate
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        className="area-5"
      >
        <ul
          className="d-flex flex-row justify-content-between list-unstyled max-width-222 p-0 mb-4"
        >
          <li>
            <a
              href="https://www.facebook.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Facebook"
            >
              <span
                aria-hidden={true}
                className="fa fa-facebook-square fa-2x"
                id="edx-footer-icon-facebook"
              />
              <span
                className="sr-only"
              >
                Like Open Edx on Facebook
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.twitter.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Twitter"
            >
              <span
                aria-hidden={true}
                className="fa fa-twitter-square fa-2x"
                id="edx-footer-icon-twitter"
              />
              <span
                className="sr-only"
              >
                Follow Open Edx on Twitter
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.youtube.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Youtube"
            >
              <span
                aria-hidden={true}
                className="fa fa-youtube-square fa-2x"
                id="edx-footer-icon-youtube"
              />
              <span
                className="sr-only"
              >
                Subscribe to the Open Edx YouTube channel
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.linkedin.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="LinkedIn"
            >
              <span
                aria-hidden={true}
                className="fa fa-linkedin-square fa-2x"
                id="edx-footer-icon-linkedin"
              />
              <span
                className="sr-only"
              >
                Follow Open Edx on LinkedIn
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://plus.google.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Google+"
            >
              <span
                aria-hidden={true}
                className="fa fa-google-plus-square fa-2x"
                id="edx-footer-icon-google"
              />
              <span
                className="sr-only"
              >
                Follow Open Edx on Google+
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://reddit.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Reddit"
            >
              <span
                aria-hidden={true}
                className="fa fa-reddit-square fa-2x"
                id="edx-footer-icon-reddit"
              />
              <span
                className="sr-only"
              >
                Subscribe to the Open Edx subreddit
              </span>
            </a>
          </li>
        </ul>
        <ul
          className="d-flex flex-row justify-content-between list-unstyled max-width-264 p-0 mb-5"
        >
          <li>
            <a
              href="https://store.apple.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
            >
              <img
                alt="Download the Open Edx mobile app from the Apple App Store"
                className="max-height-39"
                src="https://prod-edxapp.edx-cdn.org/static/images/app/app_store_badge_135x40.d0558d910630.svg"
              />
            </a>
          </li>
          <li>
            <a
              href="https://play.google.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
            >
              <img
                alt="Download the Open Edx mobile app from Google Play"
                className="max-height-39"
                src="https://prod-edxapp.edx-cdn.org/static/images/app/google_play_badge_45.6ea466e328da.png"
              />
            </a>
          </li>
        </ul>
        <p>
          <span>
            © 2012–2019 Open Edx Inc.
          </span>
          <br />
          <span>
            EdX, Open edX, and MicroMasters are registered trademarks of edX Inc. | 粤ICP备17044299号-2
          </span>
        </p>
      </div>
    </div>
  </footer>,
]
`;

exports[`ProgramListPage renders fetching program error page when there are issues fetching the user programs 1`] = `
Array [
  <div
    className="position-absolute"
  >
    <a
      className="skip-nav-link sr-only sr-only-focusable btn btn-primary px-2 py-1 mt-3 ml-2"
      href="#content"
    >
      Skip to main content
    </a>
  </div>,
  <div
    className="container my-4"
  >
    <div
      className="alert fade alert-danger show"
      hidden={false}
      role="alert"
    >
      <div
        className="alert-dialog"
      >
        <div
          className="d-flex"
        >
          <div>
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-exclamation-triangle fa-w-18 mr-2"
              data-icon="exclamation-triangle"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </div>
          <div>
            An error occurred while fetching your program enrollments. Please try again later.
          </div>
        </div>
      </div>
    </div>
  </div>,
  <footer
    aria-label="Page Footer"
    className="footer d-flex justify-content-center border-top py-3 px-4"
    role="contentinfo"
  >
    <div
      className="max-width-1180 d-grid"
    >
      <div
        className="area-1"
      >
        <a
          aria-label="Open Edx Home"
          href="https://www.example.com/"
          onClick={[Function]}
          target="_self"
        >
          <img
            alt="Open Edx logo"
            src={
              Object {
                "process": [Function],
              }
            }
          />
        </a>
      </div>
      <div
        className="area-2"
      >
        <h2>
          Open Edx
        </h2>
        <ul
          className="list-unstyled p-0 m-0"
        >
          <li>
            <a
              href="https://www.example.com/about-us"
            >
              <span>
                About
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/enterprise"
            >
              <span>
                Open Edx for Business
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/affiliate-program"
            >
              <span>
                Affiliates
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/open"
            >
              <span>
                Open Open Edx
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/careers"
            >
              <span>
                Careers
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/news-announcements"
            >
              <span>
                News
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        className="area-3"
      >
        <h2>
          <span>
            Legal
          </span>
        </h2>
        <ul
          className="list-unstyled p-0 m-0"
        >
          <li>
            <a
              href="https://www.example.com/terms-of-service"
            >
              <span>
                Terms of Service & Honor Code
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/privacy-policy"
            >
              <span>
                Privacy Policy
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/accessibility"
            >
              <span>
                Accessibility Policy
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/trademarks"
            >
              <span>
                Trademark Policy
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/sitemap"
            >
              <span>
                Sitemap
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        className="area-4"
      >
        <h2>
          <span>
            Connect
          </span>
        </h2>
        <ul
          className="list-unstyled p-0 m-0"
        >
          <li>
            <a
              href="https://www.example.com/blog"
            >
              <span>
                Blog
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/contact"
            >
              <span>
                Contact Us
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/support"
            >
              <span>
                Help Center
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/media-kit"
            >
              <span>
                Media Kit
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.example.com/donate"
            >
              <span>
                Donate
              </span>
            </a>
          </li>
        </ul>
      </div>
      <div
        className="area-5"
      >
        <ul
          className="d-flex flex-row justify-content-between list-unstyled max-width-222 p-0 mb-4"
        >
          <li>
            <a
              href="https://www.facebook.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Facebook"
            >
              <span
                aria-hidden={true}
                className="fa fa-facebook-square fa-2x"
                id="edx-footer-icon-facebook"
              />
              <span
                className="sr-only"
              >
                Like Open Edx on Facebook
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.twitter.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Twitter"
            >
              <span
                aria-hidden={true}
                className="fa fa-twitter-square fa-2x"
                id="edx-footer-icon-twitter"
              />
              <span
                className="sr-only"
              >
                Follow Open Edx on Twitter
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.youtube.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Youtube"
            >
              <span
                aria-hidden={true}
                className="fa fa-youtube-square fa-2x"
                id="edx-footer-icon-youtube"
              />
              <span
                className="sr-only"
              >
                Subscribe to the Open Edx YouTube channel
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://www.linkedin.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="LinkedIn"
            >
              <span
                aria-hidden={true}
                className="fa fa-linkedin-square fa-2x"
                id="edx-footer-icon-linkedin"
              />
              <span
                className="sr-only"
              >
                Follow Open Edx on LinkedIn
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://plus.google.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Google+"
            >
              <span
                aria-hidden={true}
                className="fa fa-google-plus-square fa-2x"
                id="edx-footer-icon-google"
              />
              <span
                className="sr-only"
              >
                Follow Open Edx on Google+
              </span>
            </a>
          </li>
          <li>
            <a
              href="https://reddit.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
              title="Reddit"
            >
              <span
                aria-hidden={true}
                className="fa fa-reddit-square fa-2x"
                id="edx-footer-icon-reddit"
              />
              <span
                className="sr-only"
              >
                Subscribe to the Open Edx subreddit
              </span>
            </a>
          </li>
        </ul>
        <ul
          className="d-flex flex-row justify-content-between list-unstyled max-width-264 p-0 mb-5"
        >
          <li>
            <a
              href="https://store.apple.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
            >
              <img
                alt="Download the Open Edx mobile app from the Apple App Store"
                className="max-height-39"
                src="https://prod-edxapp.edx-cdn.org/static/images/app/app_store_badge_135x40.d0558d910630.svg"
              />
            </a>
          </li>
          <li>
            <a
              href="https://play.google.com"
              onClick={[Function]}
              rel="noopener noreferrer"
              target="_blank"
            >
              <img
                alt="Download the Open Edx mobile app from Google Play"
                className="max-height-39"
                src="https://prod-edxapp.edx-cdn.org/static/images/app/google_play_badge_45.6ea466e328da.png"
              />
            </a>
          </li>
        </ul>
        <p>
          <span>
            © 2012–2019 Open Edx Inc.
          </span>
          <br />
          <span>
            EdX, Open edX, and MicroMasters are registered trademarks of edX Inc. | 粤ICP备17044299号-2
          </span>
        </p>
      </div>
    </div>
  </footer>,
]
`;
